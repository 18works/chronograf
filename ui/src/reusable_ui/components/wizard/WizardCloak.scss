@import 'src/style/modules/influx-colors';
@import 'src/style/modules/variables';

.step-title {
  text-align: center;
  font-size: $ix-text-base-2;
}