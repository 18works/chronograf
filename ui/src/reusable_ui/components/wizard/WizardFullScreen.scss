@import 'src/style/modules/influx-colors';
@import 'src/style/modules/variables';
@import 'src/style/modules/mixins';

.wizard-container {
  padding: 30px;
  display: flex;
  border-radius: $radius;
  max-height: 60vh;
  min-width: 800px;
  background-color: $g1-raven;
}

.wizard-title {
  margin-top: $ix-marg-c;
}