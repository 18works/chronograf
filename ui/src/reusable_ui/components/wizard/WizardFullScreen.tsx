import React, {PureComponent, ReactNode} from 'react'

// import {} from 'src/types'

interface Props {
  children: ReactNode
}

class WizardFullScreen extends PureComponent<Props> {
  public render() {
    const {children} = this.props

    return <div className="progress-bar">Step: {children}</div>
  }
}

export default WizardFullScreen
