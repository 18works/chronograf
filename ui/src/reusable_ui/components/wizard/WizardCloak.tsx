import React, {PureComponent, ReactElement} from 'react'
import WizardProgressBar from 'src/reusable_ui/components/wizard/WizardProgressBar'

import {WizardStepProps, StepStatus, Step} from 'src/types/wizard'

import 'src/reusable_ui/components/wizard/WizardCloak.scss'

interface State {
  steps: Step[]
  currentStepIndex: number
}

interface Props {
  children: Array<ReactElement<WizardStepProps>>
  handleSkip?: () => void
  skipLinkText?: string
}

class WizardCloak extends PureComponent<Props, State> {
  public static getDerivedStateFromProps(props: Props, state: State) {
    let {currentStepIndex} = state
    const childSteps = React.Children.map(
      props.children,
      (child: ReactElement<WizardStepProps>, i) => {
        const isComplete = child.props.isComplete()
        if (currentStepIndex === -1 && !isComplete) {
          currentStepIndex = i
        }
        return {
          title: child.props.title,
          stepStatus: isComplete ? StepStatus.Complete : StepStatus.Incomplete,
        }
      }
    )
    if (currentStepIndex === -1) {
      currentStepIndex = childSteps.length - 1
    }
    return {steps: childSteps, currentStepIndex}
  }

  constructor(props: Props) {
    super(props)
    this.state = {
      steps: [],
      currentStepIndex: -1,
    }
  }

  public render() {
    const {steps, currentStepIndex} = this.state
    const {skipLinkText, handleSkip} = this.props

    return (
      <div className="wizard-cloak">
        <div className="progress-header">
          <h2 className="step-title">{this.CurrentChild.props.title}</h2>
          <WizardProgressBar
            handleJump={this.jumpToStep}
            steps={steps}
            currentStepIndex={currentStepIndex}
          />
          {this.tipText}
        </div>
        {this.CurrentChild}
        <p className="skip-link">
          <a onClick={handleSkip}> {skipLinkText || 'skip'}</a>
        </p>
      </div>
    )
  }

  private incrementStep = () => {
    this.setState(prevState => {
      return {
        currentStepIndex: prevState.currentStepIndex + 1,
      }
    })
  }

  private decrementStep = () => {
    this.setState(prevState => {
      return {
        currentStepIndex: prevState.currentStepIndex - 1,
      }
    })
  }

  private jumpToStep = jumpIndex => () => {
    this.setState({
      currentStepIndex: jumpIndex,
    })
  }

  private get CurrentChild(): JSX.Element {
    const {children, handleSkip} = this.props
    const {currentStepIndex, steps} = this.state
    const lastStep = currentStepIndex === steps.length - 1

    const advance = lastStep ? handleSkip : this.incrementStep

    const retreat = currentStepIndex === 0 ? null : this.decrementStep

    return React.cloneElement<WizardStepProps>(children[currentStepIndex], {
      increment: advance,
      decrement: retreat,
      lastStep,
    })
  }

  private get tipText() {
    const {currentStepIndex} = this.state
    const {
      props: {tipText},
    } = this.props.children[currentStepIndex]

    if (tipText) {
      return (
        <div className="tip-text">
          <p>{tipText}</p>
        </div>
      )
    }
    return null
  }
}

export default WizardCloak
