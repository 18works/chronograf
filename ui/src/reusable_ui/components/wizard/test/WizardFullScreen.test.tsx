import React from 'react'
import {shallow} from 'enzyme'

import WizardFullScreen from 'src/reusable_ui/components/wizard/WizardFullScreen'
import WizardCloak from 'src/reusable_ui/components/wizard/WizardCloak'
import SplashPage from 'src/shared/components/SplashPage'

describe('WizardFullScreen', () => {
  let wrapper

  const wrapperSetup = (override = {}) => {
    const props = {
      children: null,
      title: undefined,
      skipLinkText: undefined,
      handleSkip: undefined,
      ...override,
    }

    return shallow(<WizardFullScreen {...props} />)
  }

  beforeEach(() => {
    jest.resetAllMocks()
    wrapper = wrapperSetup()
  })

  it('mounts without exploding', () => {
    expect(wrapper).toHaveLength(1)
  })

  it('renders one SplashPage component', () => {
    expect(wrapper.find(SplashPage)).toHaveLength(1)
  })

  it('renders no WizardCloak component', () => {
    expect(wrapper.find(WizardCloak)).toHaveLength(0)
  })

  it('matches snapshot with minimal props', () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe('with children', () => {
    it('renders one WizardCloak component', () => {
      wrapper = wrapperSetup({children: {}})
      expect(wrapper.find(WizardCloak)).toHaveLength(1)
    })

    it('matches snapshot with children props', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })
})
