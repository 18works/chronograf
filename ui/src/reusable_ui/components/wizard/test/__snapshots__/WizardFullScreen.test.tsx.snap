// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WizardFullScreen matches snapshot with minimal props 1`] = `
<SplashPage>
  <React.Fragment>
    <h1
      className="wizard-title"
    />
    <div
      className="wizard-container"
    />
  </React.Fragment>
</SplashPage>
`;

exports[`WizardFullScreen with children matches snapshot with children props 1`] = `
<SplashPage>
  <React.Fragment>
    <h1
      className="wizard-title"
    />
    <div
      className="wizard-container"
    />
  </React.Fragment>
</SplashPage>
`;
